File: apollo-normalized-cache/src/main/java/com/apollographql/apollo/cache/normalized/CacheReference.java
package com.apollographql.apollo.cache.normalized;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class CacheReference {

  private final String key;
  private static final Pattern SERIALIZATION_REGEX_PATTERN = Pattern.compile("ApolloCacheReference\\{(.*)\\}");
  private static final String SERIALIZATION_TEMPLATE = "ApolloCacheReference{%s}";

  public CacheReference(String key) {
    this.key = key;
  }

  public String key() {
    return key;
  }

  @Override public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || getClass() != o.getClass()) return false;

    CacheReference that = (CacheReference) o;

    return key != null ? key.equals(that.key) : that.key == null;

  }

  @Override public int hashCode() {
    return key != null ? key.hashCode() : 0;
  }

  @Override public String toString() {
    return key;
  }

  public String serialize() {
    return String.format(SERIALIZATION_TEMPLATE, key);
  }

  public static CacheReference deserialize(String serializedCacheReference) {
    Matcher matcher = SERIALIZATION_REGEX_PATTERN.matcher(serializedCacheReference);
    if (!matcher.find() || matcher.groupCount() != 1) {
      throw new IllegalArgumentException("Not a cache reference: " + serializedCacheReference
          + " Must be of the form:" + SERIALIZATION_TEMPLATE);
    }
    return new CacheReference(matcher.group(1));
  }

  public static boolean canDeserialize(String value) {
    Matcher matcher = SERIALIZATION_REGEX_PATTERN.matcher(value);
    return matcher.matches();
  }

}

File: apollo-normalized-cache-api/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/CacheReference.kt
package com.apollographql.apollo3.cache.normalized

import kotlin.jvm.JvmStatic

class CacheReference(val key: String) {

  override fun equals(other: Any?): Boolean {
    return key == (other as? CacheReference)?.key
  }

  override fun hashCode(): Int = key.hashCode()

  override fun toString(): String = key

  fun serialize(): String {
    return "$SERIALIZATION_TEMPLATE{$key}"
  }

  companion object {
    private val SERIALIZATION_REGEX_PATTERN = Regex("ApolloCacheReference\\{(.*)\\}")
    private const val SERIALIZATION_TEMPLATE = "ApolloCacheReference"

    @JvmStatic
    fun deserialize(serializedCacheReference: String): CacheReference {
      val values = SERIALIZATION_REGEX_PATTERN.matchEntire(serializedCacheReference)?.groupValues
      require(values != null && values.size > 1) {
        "Not a cache reference: $serializedCacheReference Must be of the form: $SERIALIZATION_TEMPLATE{%s}"
      }
      return CacheReference(values[1])
    }

    @JvmStatic
    fun canDeserialize(value: String): Boolean {
      return SERIALIZATION_REGEX_PATTERN.matches(value)
    }
  }

}
